### ISSUE TITLE
`children` can return unexpected results for ambiguous taxon names

### ISSUE THREAD

**Issue by arendsee** [ropensci/taxize#647]
<details> <summary><strong>Session Info</strong></summary>

```r
Session info ------------------------------------------------------------------
 setting  value
 version  R version 3.4.2 (2017-09-28)
 system   x86_64, linux-gnu
 ui       X11
 language
 collate  en_US.UTF-8
 tz       America/Chicago
 date     2017-12-02

Packages ----------------------------------------------------------------------
 package    * version    date       source
 ape          5.0        2017-10-30 cran (@5.0)
 base       * 3.4.2      2017-11-13 local
 bold         0.5.0      2017-07-21 CRAN (R 3.4.2)
 codetools    0.2-15     2016-10-05 CRAN (R 3.4.2)
 compiler     3.4.2      2017-11-13 local
 crul         0.4.0      2017-10-02 CRAN (R 3.4.2)
 curl         3.0        2017-10-06 cran (@3.0)
 data.table   1.10.4-3   2017-10-27 cran (@1.10.4-)
 datasets   * 3.4.2      2017-11-13 local
 devtools   * 1.13.3     2017-08-02 CRAN (R 3.4.1)
 digest       0.6.12     2017-01-27 CRAN (R 3.4.1)
 foreach      1.4.3      2015-10-13 CRAN (R 3.4.2)
 graphics   * 3.4.2      2017-11-13 local
 grDevices  * 3.4.2      2017-11-13 local
 grid         3.4.2      2017-11-13 local
 httr         1.3.1      2017-08-20 CRAN (R 3.4.1)
 iterators    1.0.8      2015-10-13 CRAN (R 3.4.2)
 jsonlite     1.5        2017-06-01 CRAN (R 3.4.1)
 lattice      0.20-35    2017-03-25 CRAN (R 3.4.2)
 magrittr   * 1.5        2014-11-22 CRAN (R 3.4.1)
 memoise      1.1.0      2017-04-21 CRAN (R 3.4.1)
 methods    * 3.4.2      2017-11-13 local
 nlme         3.1-131    2017-02-06 CRAN (R 3.4.2)
 parallel     3.4.2      2017-11-13 local
 plyr         1.8.4      2016-06-08 CRAN (R 3.4.1)
 pryr       * 0.1.3      2017-10-30 cran (@0.1.3)
 R6           2.2.2      2017-06-17 CRAN (R 3.4.1)
 Rcpp         0.12.13    2017-09-28 cran (@0.12.13)
 reshape      0.8.7      2017-08-06 cran (@0.8.7)
 reshape2     1.4.2      2016-10-22 CRAN (R 3.4.1)
 stats      * 3.4.2      2017-11-13 local
 stringi      1.1.6      2017-11-17 CRAN (R 3.4.2)
 stringr      1.2.0      2017-02-18 CRAN (R 3.4.1)
 taxize     * 0.9.0.9214 2017-11-16 Github (ropensci/taxize@1e93a03)
 tools        3.4.2      2017-11-13 local
 utils      * 3.4.2      2017-11-13 local
 withr        2.0.0      2017-07-28 CRAN (R 3.4.1)
 xml2         1.1.1      2017-01-24 CRAN (R 3.4.1)
 zoo          1.8-0      2017-04-12 cran (@1.8-0)
```
</details>

For this search

``` R
library(taxize)
taxize::children(2, db='ncbi')
```

I get the results

``` R
$`2`
   childtaxa_id                  childtaxa_name childtaxa_rank
1       1936987                    Balneolaeota         phylum
2       1930617                 Calditrichaeota         phylum
3       1853220                 Rhodothermaeota         phylum
4       1802340 Nitrospinae/Tectomicrobia group        no rank
5       1783272             Terrabacteria group        no rank
6       1783270                       FCB group        no rank
7       1783257                       PVC group        no rank
8        629425                 Bacteria ferula        species
9        629405              Bacteria bahiensis        species
10       629404                Bacteria baculus        species
11       629403              Bacteria apolinari        species
12       629401                Bacteria ambigua        species
13       629398       Bacteria acuminatocercata        species
14       629397              Bacteria aborigena        species
15       629396               Bacteria abnormis        species
16       508458                   Synergistetes         phylum
17       203691                    Spirochaetes         phylum
18       200940           Thermodesulfobacteria         phylum
19       200938                  Chrysiogenetes         phylum
20       200930                 Deferribacteres         phylum
21       200918                     Thermotogae         phylum
22       200783                       Aquificae         phylum
23        74152                   Elusimicrobia         phylum
24        68297                     Dictyoglomi         phylum
25        67814                     Caldiserica         phylum
26        57723                   Acidobacteria         phylum
27        40117                     Nitrospirae         phylum
28        32066                    Fusobacteria         phylum
29         1224                  Proteobacteria         phylum

attr(,"class")
[1] "children"
attr(,"db")
[1] "ncbi"
```

The NCBI taxon '2' represents the Bacteria domain. As expected, I get all the bacterial phyla. However, I also get a bunch of insects from the genus 'Bacteria'. It seems that the taxon ID is being converted to the name and then the name is being searched. This isn't safe since taxon names are not unique (unfortunately).

Another example is the genus 'Brunia', which is both a plant and a moth genus. For this name, `children` returns a mixture of the moth (e.g. *Brunia antica*) and plant (e.g. *Brunia sphaerocephala*) species:

``` R
> taxize::children('Brunia', db='ncbi')
$Brunia
   childtaxa_id        childtaxa_name childtaxa_rank
   1       1817978       Brunia sarawaca        species
   2       1742571                Brunia       subgenus
   3       1426864       Brunia dorsalis        species
   4       1266089        Brunia ekeikei        species
   5        938262         Brunia antica        species
   6        863415         Brunia laevis        species
   7        266454    Brunia noduliflora        species
   8        266453       Brunia neglecta        species
   9        266452   Brunia macrocephala        species
   10       266451       Brunia powrieae        species
   11       266450   Brunia fragarioides        species
   12       266449       Brunia paleacea        species
   13       266448 Brunia sphaerocephala        species
   14       266447  Brunia latebracteata        species
   15       266442         Pseudobaeckea       subgenus
   16       266430              Raspalia       subgenus
   17       266427           Mniothamnea       subgenus
   18       266421           Lonchostoma       subgenus
```

Though, oddly, in this case, I do get the correct results when using the taxon IDs: 

``` R
taxize::children(57717, db='ncbi')
```

**Comment by sckott** [comment]
I think when names are given (e.g., `Brunia`) we're sort of stuck with that problem. Though if there's a way to give a higher rank to eliminate responses not wanted that would be good.

With IDs though I think we should be able to fix that problem. Looking into what's going on with `taxize::children(2, db='ncbi')`

**Comment by sckott** [comment]
@zachary-foster It does seem that internally within `ncbi_children` even if an ID is given the function gets a taxonomic name and gets children based on the name instead of the ID. Is it possible to instead use the ID?

**Comment by zachary-foster** [comment]
Hi @sckott. Sorry for the delay. I just saw this. First of all, who thought it was a good idea to name an insect genus "Bacteria"?! Oh well. 

I tried a few things and I cant find a way to use the ID instead of the name. The `esearch` API has a `uid` field, but the part that looks for children is in the `term` field and I dont see how to specify that it is a uid AND I want children. Here is what that search looks like: 

```
https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=taxonomy&term=Bacteria[Next+Level]+AND+cellular+organisms[subtree]&RetMax=1000&RetStart=0
```

We would need something like `term=2[uid][Next+Level]`, but that does not work. `term=2[Next+Level]` does not work either. `term=2[uid]` is valid, but does not return the children. Maybe we can use something besides `esearch`, but I don't know what. 

I might look into this more when I get a chance, but for now I dont have any leads.

**Comment by arendsee** [comment]
@zachary-foster Oddly, the Bacteria genus of walking sticks was named first.

Yeah, entrez is a tar pit. Getting the immediate children of a node in a tree should be trivial. I've spent the last 2 hours messing with this and have ended up in exactly the same place as you.

Perhaps we can use entrez `elink` instead of `esearch`. I can get all the taxa that have the query ID in their lineage. Using `rentrez`:

```R
rentrez::entrez_link(
    db     = 'taxonomy',
    dbfrom = 'taxonomy',
    id     = ids
)$links$taxonomy_taxonomy_child
``` 

Maybe the command could be filtered in some way to just keep immediate children ...

The SQL operation is pretty trivial:

```SQL
SELECT tax_id FROM nodes WHERE parent_tax_id == 2;
```

I might throw this question at biostar or the NCBI help desk.


Aside, I think you can simplify the search to:

```
https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=taxonomy&term=Bacteria[Next+Level]&RetMax=1000&RetStart=0
```

**Comment by zachary-foster** [comment]
> Oddly, the Bacteria genus of walking sticks was named first.

Interesting! I guess thats why they have not changed it.

>Yeah, entrez is a tar pit. Getting the immediate children of a node in a tree should be trivial. I've spent the last 2 hours messing with this and have ended up in exactly the same place as you.

Haha, tar pit is right: messy and full of fossils. 

> Maybe the command could be filtered in some way to just keep immediate children ...

elink looks like something to explore more. I ran the rentrez command and it took a while to return, so I am concerned that even if we could filter for immediate children, it might be too slow to be practical.

> I might throw this question at biostar or the NCBI help desk.

Cool!

>Aside, I think you can simplify the search to:

Thanks!

**Comment by arendsee** [comment]
OK, I've posted a question on this topic on Biostars:

https://www.biostars.org/p/288432/

**Comment by zachary-foster** [comment]
This looks promising, although it requires both the ID and name to work: 

```
https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=taxonomy&db=taxonomy&id=2&term=Bacteria[Next+Level]
```

I looked up about half the IDs it returns and they were all bacterial phyla

**Comment by zachary-foster** [comment]
Its not too fast, but it works: 

```
> children(2, db='ncbi')
No ENTREZ API key provided
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/
No ENTREZ API key provided
See https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/
$`2`
   childtaxa_id        childtaxa_name childtaxa_rank
1        508458         Synergistetes         phylum
2        203691          Spirochaetes         phylum
3        200940 Thermodesulfobacteria         phylum
4        200938        Chrysiogenetes         phylum
5        200930       Deferribacteres         phylum
6        200918           Thermotogae         phylum
7        200783             Aquificae         phylum
8         74152         Elusimicrobia         phylum
9         68297           Dictyoglomi         phylum
10        67814           Caldiserica         phylum
11        57723         Acidobacteria         phylum
12        40117           Nitrospirae         phylum
13        32066          Fusobacteria         phylum
14         1224        Proteobacteria         phylum

attr(,"class")
[1] "children"
attr(,"db")
[1] "ncbi"
```

Still need to make sure I did not break anything else.

**Comment by zachary-foster** [comment]
@sckott, It looks like you had addresses this issue with the "ancestor" option? 

https://github.com/ropensci/taxize/blob/b1e9c731cf987c0e737ffc82d244b1ecad0f236c/R/ncbi_children.R#L39-L45

Did that work? Or did it only work for names, and not IDs?

**Comment by sckott** [comment]
@zachary-foster yeah, that's what the `ancestor` param was aiming to do

**Comment by sckott** [comment]
closing, should be resolved by #659 - @arendsee let us know if not resolved
